pages/advanced/smart-account-modules/smart-account-modules-tutorial.mdx

```diff
@@ -0,0 +1,964 @@
+
+`TokenTransferModule.sol` will contain the module code which will be used further in the tutorial for interacting with the UI.
+
+#### Step 1. Create empty contract
+
+```solidity
+// SPDX-License-Identifier: LGPL-3.0
+pragma solidity ^0.8.0;
+// Imports will be added here
+
+contract TokenTransferModule {
+  // State variables will be added here
+
+  // Constructor will be added here
+
+  // Functions will be added here
+}
+```
+
+Explanation:
+- **SPDX License Identifier**: Specifies the license type.
+- **`pragma solidity ^0.8.0`**: Defines the Solidity compiler version.
+- **`contract TokenWithdrawModule`**: Declares the contract name.
+
+#### Step 2: Import required dependencies
+
+
+```solidity
+import "@safe-global/safe-contracts/contracts/common/Enum.sol";
+import "@safe-global/safe-contracts/contracts/Safe.sol";
+```
+
+Explanation:
+- **`Enum.sol`**: Provides Enum `Operation` which can have values like `Call` or `DelegateCall`. This will be used further in the contract when a module calls a Safe account where the module specifies the operation type.
+- **`Safe.sol`**: Includes the Safe contract interface to interact with Safe accounts.
+
+#### Step 3: Define state variables
+
+Declare the necessary state variables for the contract.
+
+```solidity
+bytes32 public immutable PERMIT_TYPEHASH =
+    keccak256(
+        "TokenTransfer(uint256 amount, address _beneficiary, uint256 nonce, uint256 deadline)"
+    );
+address public immutable safeAddress;
+address public immutable tokenAddress;
+mapping(address => uint256) public nonces;
+```
+
+Explanation:
+- **`PERMIT_TYPEHASH`**: Used to construct the signature hash for the token transfer. 
+- **`safeAddress`**: Stores the Safe contract address.
+- **`tokenAddress`**: Stores the ERC20 token contract address.
+- **`nonces`**: Tracks unique nonce to prevent replay attacks.
+
+#### Step 4: Create the Constructor
+
+Define a constructor to initialize the Safe and token contract addresses.
+
+```solidity
+constructor(address _tokenAddress, address _safeAddress) {
+    tokenAddress = _tokenAddress;
+    safeAddress = _safeAddress;
+}
+```
+
+- Initializes `tokenAddress` and `safeAddress` with provided values during deployment. Thus, in this module the token and Safe addresses are fixed.
+
+#### Step 5: Implement the `tokenTransfer` function
+
+Add a function to handle token transfers from the Safe.
+
+```solidity
+function tokenTransfer(
+    uint _amount,
+    address _beneficiary,
+    uint256 _deadline,
+    bytes memory _signatures
+) public {
+    require(_deadline >= block.timestamp, "expired deadline");
+
+    bytes32 signatureData = keccak256(
+        abi.encode(
+            PERMIT_TYPEHASH,
+            _amount,
+            msg.sender,
+            nonces[msg.sender]++,
+            _deadline
+        )
+    );
+
+    bytes32 hash = keccak256(
+        abi.encodePacked("\x19\x01", getDomainSeparator(), signatureData)
```

Comment:
`getDomainSeparator` is used here but only added in the next step. I'd define everything that's needed for signing first